File: apps/server/openhive/auth/providers.py
"""Provider registry — static metadata for each OAuth provider we support.

Flow implementations live in auth.services.*. Each entry defines how the browser-side
UI should render the provider (label, description) and how the backend should talk to
its OAuth endpoints.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


Kind = Literal["auth_code", "device_code"]


@dataclass(frozen=True)
class ProviderDef:
    id: str
    label: str
    kind: Kind
    description: str


# Order matters — this is the display order in the UI.
PROVIDERS: list[ProviderDef] = [
    ProviderDef(
        id="claude-code",
        label="Claude Code",
        kind="auth_code",
        description="Use your Claude Code subscription to power agents with Claude models.",
    ),
    ProviderDef(
        id="codex",
        label="OpenAI Codex",
        kind="auth_code",
        description="Use your Codex (ChatGPT) subscription for agents running on GPT models.",
    ),
    ProviderDef(
        id="copilot",
        label="GitHub Copilot",
        kind="device_code",
        description="Use your GitHub Copilot subscription. Activates via device-code login.",
    ),
]

_BY_ID = {p.id: p for p in PROVIDERS}


def get(provider_id: str) -> ProviderDef | None:
    return _BY_ID.get(provider_id)

File: apps/web/lib/server/auth/providers.ts
/**
 * Provider registry. Ports apps/server/openhive/auth/providers.py.
 * Static metadata — flow implementations live in ./claude, ./codex, ./copilot.
 */

export type FlowKind = 'auth_code' | 'device_code' | 'api_key'

export interface ProviderDef {
  id: string
  label: string
  kind: FlowKind
  description: string
}

/** Display order in the UI. */
export const PROVIDERS: ProviderDef[] = [
  {
    id: 'codex',
    label: 'OpenAI Codex',
    kind: 'auth_code',
    description:
      'Use your Codex (ChatGPT) subscription for agents running on GPT models.',
  },
  {
    id: 'copilot',
    label: 'GitHub Copilot',
    kind: 'device_code',
    description:
      'Use your GitHub Copilot subscription. Activates via device-code login.',
  },
  {
    id: 'openai',
    label: 'OpenAI',
    kind: 'api_key',
    description: 'Direct OpenAI API access via API key (platform.openai.com).',
  },
  {
    id: 'gemini',
    label: 'Google Gemini',
    kind: 'api_key',
    description: 'Gemini API access via API key (aistudio.google.com).',
  },
  {
    id: 'vertex-ai',
    label: 'Vertex AI',
    kind: 'api_key',
    description: 'Google Cloud Vertex AI via service-account JSON.',
  },
]

const BY_ID = new Map<string, ProviderDef>(PROVIDERS.map((p) => [p.id, p]))

export function getProvider(id: string): ProviderDef | null {
  return BY_ID.get(id) ?? null
}
